### PR TITLE
[Docs] Add docs for Sky Serve

### PR DESCRIPTION
**Pull request by cblmemo**
Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by concretevitamin — COMMENTED** [review]
Nice!

**Comment by concretevitamin** [comment]
We also may want to replace some snippets (e.g., `sky serve status`?) with screenshots which look way better with colors than the current formattin.

**Review by landscapepainter — COMMENTED** [review]
Had a chance to go thorugh the doc to learn how SkyServe works. Left some comments on what I found.

File: docs/source/examples/sky-serve.rst [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    </head>
+    <body>
+        <p>Hello, Sky Serve!</p>
+    </body>
+    </html>
+
+.. note::
+
+  The :code:`curl` command won't follow the redirect and print the content of the redirected page by default. Since we are using HTTP-redirect, you need to use :code:`curl -L <endpoint-url>`.
+
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply docs/source/examples/sky-serve.rst:147]: If we are adding this, do you think we still need the vicuna example? Not sure if it is a little bit redundant if we include TGI...

File: docs/source/index.rst [hunk 2]
```diff
@@ -110,6 +110,13 @@ Documentation
    reference/kubernetes/index
    running-jobs/index
 
+.. toctree::
+   :maxdepth: 1
+   :caption: Multi-Cloud Serving
+
+   examples/sky-serve
+   examples/service-yaml-spec
```
Comment (concretevitamin) [inline docs/source/index.rst:118]: We probably want to move these to `serving/*`. This will show up in URLs.

File: docs/source/index.rst [hunk 3]
```diff
@@ -110,6 +110,13 @@ Documentation
    reference/kubernetes/index
    running-jobs/index
 
+.. toctree::
+   :maxdepth: 1
+   :caption: Multi-Cloud Serving
```
Comment (concretevitamin) [inline docs/source/index.rst:115]:
```suggestion
   :caption: SkyServe: Model Serving
```

File: docs/source/examples/sky-serve.rst [hunk 4]
```diff
@@ -0,0 +1,333 @@
+.. _sky-serve:
+
+Sky Serve
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:3]: Let's change all "Sky Serve" to "SkyServe" in both files?

File: docs/source/examples/sky-serve.rst [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+=========
+
+Sky Serve is SkyPilot's serving library. Sky Serve takes an existing serving
+framework and deploys it across one or more regions or clouds.
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:7]:
```suggestion
SkyServe (short for SkyPilot Serving) takes an existing serving
framework and deploys it across one or more regions or clouds.
```

File: docs/source/examples/service-yaml-spec.rst [hunk 6]
```diff
@@ -33,7 +33,9 @@ Available fields:
       # and the default initial delay, you can use the following syntax:
       readiness_probe: /v1/models
 
-      # Replica autoscaling policy (required). This describe how SkyServe autoscale
+      # One of the two following fields (replica_policy or replicas) is required.
+
+      # Replica autoscaling policy. This describe how SkyServe autoscale
```
Comment (concretevitamin) [inline docs/source/examples/service-yaml-spec.rst:38]:
```suggestion
      # Replica autoscaling policy. This describes how SkyServe autoscales
```

File: docs/source/examples/sky-serve.rst [hunk 7]
```diff
@@ -3,20 +3,44 @@
...
+
+.. * Serve on scarce resources (e.g., A100; spot) with **reduced costs and increased availability**
 
 Why Sky Serve?
 
-* Allocate scarce resources (e.g., A100) **across regions and clouds**
-* Autoscale your endpoint deployment with load balancing
-* Manage your multi-cloud resources with a single control plane
+* **Bring any serving framework** (vLLM, TGI, FastAPI, ...) and scale it across regions/clouds
+* **Reduce costs and increase availability** of service replicas by leveraging multiple/cheaper locations and hardware
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:14]:
```suggestion
* **Reduce costs and increase availability** of service replicas by leveraging multiple/cheaper locations and hardware (spot instances)
```

File: docs/source/examples/sky-serve.rst [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+Sky Serve has a centralized controller VM that manages the deployment of your service. Each service will have a process group to manage its replicas and route traffic to them.
+
+It is composed of the following components:
+
+#. **Controller**: The controller will monitor the status of the replicas and re-launch a new replica if one of them fails. It also autoscales the number of replicas if autoscaling config is set (see :ref:`Service YAML spec <service-yaml-spec>` for more information).
+#. **Load Balancer**: The load balancer will route the traffic to all ready replicas. It is a lightweight HTTP server that listens on the service endpoint and **HTTP-redirects** the requests to one of the replicas.
+
+All of the process group shares a single controller VM. The controller VM will be launched in the cloud with the best price/performance ratio. You can also :ref:`customize the controller resources <customizing-sky-serve-controller-resources>` based on your needs.
+
+An end-to-end LLM example
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:153]:
```suggestion
Tutorial: Serve a Chatbot LLM!
```

File: docs/source/examples/sky-serve.rst [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+-------------------------
+
+Below we show an end-to-end example of deploying a LLM model with Sky Serve. We'll use the `Vicuna OpenAI API Endpoint YAML <https://github.com/skypilot-org/skypilot/blob/master/llm/vicuna/serve-openai-api-endpoint.yaml>`_ as an example:
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:156]:
```suggestion
Let’s bring up a real LLM chat service with FastChat + Vicuna. We'll use the `Vicuna OpenAI API Endpoint YAML <https://github.com/skypilot-org/skypilot/blob/master/llm/vicuna/serve-openai-api-endpoint.yaml>`_ as an example:
```

File: docs/source/examples/sky-serve.rst [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+Under the hood, :code:`sky serve up`:
+
+#. Launches a controller which handles autoscaling, monitoring and load balancing;
+#. Returns a Service Endpoint which will be used to accept traffic;
+#. Meanwhile, the controller provisions replica VMs which later run the services;
+#. Once any replica is ready, the requests sent to the Service Endpoint will be **HTTP-redirect** to one of the endpoint replicas.
+
+After the controller is provisioned, you'll see the following in :code:`sky serve status` output:
+
+.. image:: ../images/sky-serve-status-output-provisioning.png
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:113]: One problem to fix throughout this file: the para immediately after an image becomes too hard to read due to no vertical space.

File: docs/source/examples/sky-serve.rst [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+
+    # Here goes other task config
+
+Now you have a Service YAML that can be used with Sky Serve! Simply run :code:`sky serve up vicuna.yaml -n vicuna` to deploy the service (use :code:`-n` to give your service a name!). After a while, there will be an OpenAI Compatible API endpoint ready to accept traffic (:code:`44.201.113.28:30001` in the following example):
+
+.. image:: ../images/sky-serve-status-vicuna-ready.png
+    :width: 800
+    :align: center
+    :alt: sky-serve-status-vicuna-ready
+
```
Comment (concretevitamin) [inline docs/source/examples/sky-serve.rst:218]: We should add a CURL based command like the user doc. Does something like this work
```
Then it’s ready to accept traffic!

$ curl -L Y.Y.Y.Y:8082/generate \
    -X POST \
    -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":20}}' \
    -H 'Content-Type: application/json'
{"generated_text":"\n nobody knows"}
```

File: docs/source/examples/sky-serve.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (concretevitamin) [reply docs/source/examples/sky-serve.rst:135]: Some redundancy is fine. The main motivation is to make the very first impression about real, useful AI serving. Currently it's HTTP server.

How about we add a "Quickstart: TGI service" section (or change it to vLLM/FastChat etc.), but keep the user doc's concise formatting -- 1 snippet showing YAML, 1 snippet showing service status, then add 1 snippet showing how to CURL it correctly. With minimal text throughout.

File: docs/source/index.rst (hunk 3 above)
Comment (cblmemo) [reply docs/source/index.rst:115]: We can do more than model serving..?

**Review by concretevitamin — APPROVED** [review]

**Pull request merged**